core/decorations.py

```diff
@@ -0,0 +1,79 @@
+from functools import reduce
+from LSP.plugin.core.css import css
+from LSP.plugin.core.protocol import Range
+from LSP.plugin.core.sessions import Session
+from LSP.plugin.core.typing import Any, List, Dict, Optional
+from LSP.plugin.core.views import range_to_region, FORMAT_MARKED_STRING, FORMAT_MARKUP_CONTENT, minihtml
+import mdpopups
+import sublime
+
+
+def publishDecorations(session: Session, decorationsParams: Any) -> None:
+    phantom_key = "metals_decoraction"
+
+    if not isinstance(decorationsParams, dict):
+        return
+    if not session:
+        return
+    uri = decorationsParams.get('uri')
+    if not uri:
+        return
+
+    session_buffer = session.get_session_buffer_for_uri_async(uri)
+    if not session_buffer:
+        return
+
+    for sv in session_buffer.session_views:
+        try:
+            phantom_set = getattr(sv, "_lsp_metals_decorations")
+        except AttributeError:
+            phantom_set = sublime.PhantomSet(sv.view, phantom_key)
+            setattr(sv, "_lsp_metals_decorations", phantom_set)
+
+        phantom_set.update(decorations_to_phantom(decorationsParams.get('options', []), sv.view))
```

Comment:
I'm pretty sure that phantoms (unlike regions) attach to a buffer so creating a separate set for each view is redundant. You can store those on the SessionBuffer.